### PR TITLE
Engine processing Tpwdcee 4470

### PR DESCRIPTION
**Pull request by tomazpu**
The PR, includes the engine processing of a response or notification over the strategy design pattern.

Please squash and merge to one commit.

### CODE REVIEW

File: wirecardpaymentgateway/controllers/front/return.php [hunk 1]
```diff
@@ -50,116 +48,55 @@
...
-            $order->setCurrentState(Configuration::get(OrderManager::WIRECARD_OS_AWAITING));
-        }
-
-        $orderPayments = OrderPayment::getByOrderReference($order->reference);
-        if (!empty($orderPayments)) {
-            $orderPayments[count($orderPayments) - 1]->transaction_id = $response->getTransactionId();
-            $orderPayments[count($orderPayments) - 1]->save();
+        if ($payment_state === self::CANCEL_PAYMENT_STATE) {
+            $response_processing = new CancelResponseProcessing();
+            $response_processing->process(null);
```
Comment (pretzelhands) [inline wirecardpaymentgateway/controllers/front/return.php]: Maybe the parameter passed to `ResponseProcessing` can be made optional? Passing in `null` explicitly feels a bit iffy.

File: wirecardpaymentgateway/controllers/front/return.php [hunk 2]
```diff
@@ -33,11 +33,9 @@
  * @license GPLv3
  */
 
-use Wirecard\PaymentSdk\Response\SuccessResponse;
-use Wirecard\PaymentSdk\Response\FailureResponse;
-use Wirecard\PaymentSdk\TransactionService;
-use Wirecard\PaymentSdk\Exception\MalformedResponseException;
-use WirecardEE\Prestashop\Helper\OrderManager;
+use WirecardEE\Prestashop\classes\EngineResponseProcessing\ReturnPaymentEngineResponseProcessing;
```
Comment (pretzelhands) [inline wirecardpaymentgateway/controllers/front/return.php]: Minor thing: Can we have `classes` in the namespace be uppercase? Inconsistency in the casing seems like a potential cause of pain in the future.

File: wirecardpaymentgateway/classes/EngineResponseProcessing/EngineResponseProcessing.php [hunk 3]
```diff
@@ -37,15 +37,16 @@
...
+ *
  * @package WirecardEE\Prestashop\classes\EngineResponseProcessing
  * @since 2.1.0
  */
 interface EngineResponseProcessing
 {
     /**
-     * @param array $response
-     * @param \WirecardPaymentGateway $module
+     * @param array|string $response
```
[the remark this answers is not in the document]
Comment (tomazpu) [reply wirecardpaymentgateway/classes/EngineResponseProcessing/EngineResponseProcessing.php]: As the return gets the values over the $_REQUEST its a array [phpdoc](https://www.php.net/manual/de/reserved.variables.request.php)
The notify gets it over php://input and is an encoded string

File: wirecardpaymentgateway/classes/EngineResponseProcessing/PaymentEngineResponseProcessing.php [hunk 4]
```diff
@@ -35,20 +35,47 @@
...
+    protected $payment;
+
+    /**
+     * @param array|string $response
+     * @param \ModuleFrontController $controller
+     * @since 2.1.0
+     */
+    public function process($response, $controller)
+    {
+        $config = $this->getPaymentConfig(
```
[the remark this answers is not in the document]
Comment (tomazpu) [reply wirecardpaymentgateway/classes/EngineResponseProcessing/PaymentEngineResponseProcessing.php]: i know that's why i left it in a method :)

File: wirecardpaymentgateway/classes/EngineResponseProcessing/PaymentEngineResponseProcessing.php [hunk 5]
```diff
@@ -35,20 +35,47 @@
...
 abstract class PaymentEngineResponseProcessing implements EngineResponseProcessing
 {
+    /** @var BackendService */
+    protected $backend_service;
+
+    /** @var Payment */
+    protected $payment;
+
+    /**
+     * @param array|string $response
```
[the remark this answers is not in the document]
Comment (tomazpu) [reply wirecardpaymentgateway/classes/EngineResponseProcessing/PaymentEngineResponseProcessing.php]: As this is the abstract method used in both notify and return i think its the correct documentation

File: wirecardpaymentgateway/controllers/front/return.php (hunk 2 above)
Comment (tomazpu) [reply wirecardpaymentgateway/controllers/front/return.php]: Is it ok if i change it in a separate task for all classes? i see the problem in the whole plugin

**Pull request merged**